File: plugin/src/main/java/org/jetbrains/iren/settings/AppSettingsComponent.java
// Copyright 2000-2020 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package org.jetbrains.iren.settings;

import com.intellij.openapi.ui.ComboBox;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.fields.IntegerField;
import com.intellij.util.ui.FormBuilder;

import javax.swing.*;
import java.time.temporal.ChronoUnit;
import java.util.Arrays;
import java.util.List;

/**
 * Supports creating and managing a {@link JPanel} for the Settings Dialog.
 */
public class AppSettingsComponent {

    private final JPanel myMainPanel;
    private final JBCheckBox myAutomaticTrainingStatus = new JBCheckBox("Automatic training of models.");
    private final IntegerField myMaxTrainingTimeText = new IntegerField("30", 0, 10000);
    private final IntegerField myVocabularyCutOff = new IntegerField("0", 0, 20);
    private final IntegerField myModelsLifetime = new IntegerField();
    private final List<ChronoUnit> chronoUnits = Arrays.asList(ChronoUnit.HOURS, ChronoUnit.DAYS, ChronoUnit.WEEKS, ChronoUnit.MONTHS);
    private final ComboBox<String> myModelsLifetimeUnit = new ComboBox<>(chronoUnits.stream().map(ChronoUnit::toString).toArray(String[]::new));

    public AppSettingsComponent() {
        int columnWidth = 9;
        myMaxTrainingTimeText.setColumns(columnWidth);
        myVocabularyCutOff.setColumns(columnWidth);
        myModelsLifetime.setColumns(columnWidth);
        myMainPanel = FormBuilder.createFormBuilder()
                .addComponent(myAutomaticTrainingStatus, 1)
                .addVerticalGap(10)
                .addLabeledComponent(new JBLabel("Maximal training time of models (s): "), myMaxTrainingTimeText, 1, false)
                .addLabeledComponent(new JBLabel("Vocabulary cutoff: "), myVocabularyCutOff, 1, false)
                .addTooltip("Remove words with small frequencies.")
                .addLabeledComponent(new JBLabel("Models lifetime: "), myModelsLifetime, 1, false)
                .addComponentToRightColumn(this.myModelsLifetimeUnit)
                .addTooltip("The time after which model will be retrained.")
                .addComponentFillVertically(new JPanel(), 0)
                .getPanel();
    }

    public JPanel getPanel() {
        return myMainPanel;
    }

    public JComponent getPreferredFocusedComponent() {
        return myAutomaticTrainingStatus;
    }

    public boolean getAutomaticTrainingStatus() {
        return myAutomaticTrainingStatus.isSelected();
    }

    public void setAutomaticTrainingStatus(boolean newStatus) {
        myAutomaticTrainingStatus.setSelected(newStatus);
    }

    public Integer getMaxTrainingTime() {
        try {
            return Integer.parseInt(myMaxTrainingTimeText.getText());
        } catch (NumberFormatException e) {
            return null;
        }
    }

    public void setMaxTrainingTime(int newText) {
        myMaxTrainingTimeText.setText(String.valueOf(newText));
    }

    public Integer getVocabularyCutOff() {
        try {
            return Integer.parseInt(myVocabularyCutOff.getText());
        } catch (NumberFormatException e) {
            return null;
        }
    }

    public void setVocabularyCutOff(int newText) {
        myVocabularyCutOff.setText(String.valueOf(newText));
    }

    public Integer getModelsLifetime() {
        try {
            return Integer.parseInt(myModelsLifetime.getText());
        } catch (NumberFormatException e) {
            return null;
        }
    }

    public void setModelsLifetime(int newText) {
        myModelsLifetime.setText(String.valueOf(newText));
    }

    public ChronoUnit getModelsLifetimeUnit() {
        return chronoUnits.get(myModelsLifetimeUnit.getSelectedIndex());
    }

    public void setModelsLifetimeUnit(ChronoUnit newUnit) {
        myModelsLifetimeUnit.setSelectedIndex(chronoUnits.indexOf(newUnit));
    }
}

File: plugin/src/main/kotlin/org/jetbrains/iren/settings/AppSettingsComponent.kt
// Copyright 2000-2020 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.jetbrains.iren.settings

import com.intellij.openapi.ui.ComboBox
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBTextField
import com.intellij.ui.dsl.builder.Cell
import com.intellij.ui.dsl.builder.columns
import com.intellij.ui.dsl.builder.panel
import com.intellij.ui.dsl.builder.selected
import org.jetbrains.iren.IRenBundle
import java.time.temporal.ChronoUnit
import javax.swing.JPanel

/**
 * Supports creating and managing a [JPanel] for the Settings Dialog.
 */
class AppSettingsComponent {
    val panel: JPanel
    private lateinit var automaticTrainingCell: Cell<JBCheckBox>
    private lateinit var maxTrainingTimeCell: Cell<JBTextField>
    private lateinit var vocabCutoffCell: Cell<JBTextField>
    private lateinit var modelsLifetimeCell: Cell<JBTextField>
    private lateinit var modelsLifetimeUnitCell: Cell<ComboBox<ChronoUnit>>
    private val chronoUnits = listOf(ChronoUnit.HOURS, ChronoUnit.DAYS, ChronoUnit.WEEKS, ChronoUnit.MONTHS)

    init {
        val columnWidth = 9
        panel = panel {
            row {
                automaticTrainingCell = checkBox(IRenBundle.message("settings.automatic.training"))
            }
            group(IRenBundle.message("settings.title.model.parameters")) {
                row(IRenBundle.message("settings.max.training.time")) {
                    maxTrainingTimeCell = intTextField(2..10000, 10).columns(columnWidth)
                    label("seconds")
                }
                row(IRenBundle.message("settings.vocabulary.cutoff")) {
                    vocabCutoffCell = intTextField(0..20, 1).columns(columnWidth)
                    comment(IRenBundle.message("settings.vocabulary.cutoff.comment"))
                }
            }
            group(IRenBundle.message("settings.title.retraining")) {
                row(IRenBundle.message("settings.models.lifetime")) {
                    modelsLifetimeCell = intTextField(1..Int.MAX_VALUE, 1).columns(columnWidth)
                    modelsLifetimeUnitCell = comboBox(chronoUnits).columns(columnWidth)
                    comment(IRenBundle.message("settings.models.lifetime.comment"))
                }.enabledIf(automaticTrainingCell.selected)
            }
        }
    }

    var automaticTrainingStatus: Boolean
        get() = automaticTrainingCell.component.isSelected
        set(newStatus) {
            automaticTrainingCell.component.isSelected = newStatus
        }

    var maxTrainingTime: Int
        get() = maxTrainingTimeCell.component.text.toInt()
        set(newText) {
            maxTrainingTimeCell.component.text = newText.toString()
        }

    var vocabularyCutOff: Int
        get() = vocabCutoffCell.component.text.toInt()
        set(newText) {
            vocabCutoffCell.component.text = newText.toString()
        }

    var modelsLifetime: Int
        get() = modelsLifetimeCell.component.text.toInt()
        set(newText) {
            modelsLifetimeCell.component.text = newText.toString()
        }
    var modelsLifetimeUnit: ChronoUnit
        get() = chronoUnits[modelsLifetimeUnitCell.component.selectedIndex]
        set(newUnit) {
            modelsLifetimeUnitCell.component.selectedIndex = chronoUnits.indexOf(newUnit)
        }
}
